### PR TITLE
[ast] Fix CDC and DFT issues

### PR DESCRIPTION
**Pull request by Jacob-Levy**
Signed-off-by: Jacob Levy <jacob.levy@nuvoton.com>

* Fix CDC & DFT Issues (ast.sv & ast_clks_byp.sv).
* Update main_pok reset through regulator SM (ast.sv & rglts_pdm_3p3v.sv).
* Update ast.hjson with CSR exclude information, access type, and different reset value for each register.
* Add flag `calibrate_usb_clk` to USB oscillator and add `%m`  to `$display` in osc files (<*>_osc.sv)

### CODE REVIEW

**Comment by Jacob-Levy** [comment]
[   legend    ]: [Q: queued, D: dispatched, P: passed, F: failed, K: killed, T: total]                                                                                        
00:13:24  [    build    ]: [Q: 000, D: 000, P: 044, F: 000, K: 000, T: 044] 100%                                                                                                        
00:34:36  [     run     ]: [Q: 000, D: 000, P: 115, F: 000, K: 000, T: 115] 100%             

CI (private) passed on local run...

**Comment by Jacob-Levy** [comment]
@matutem Added `usb_osc.sv` to this PR.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/data/ast.hjson:48]: nit: please fix alignment issue here and below:
```suggestion
        tags: [ // don't write random data to any of the AST registers
               "excl:CsrAllTests:CsrExclWrite" ],
```

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:52]: Does this really take negative numbers? I would think `urandom` implies "unsigned". I will suggest changing this to:
```systemverilog
    rand32 = $urandom_range(0, 416) * ($urandom_range(0, 1) ? 1 : -1);  // +/-416ps (+/-2% max)
```
See https://www.edaplayground.com/x/VnmG for an example.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:52]: Good catch! will fix it.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/data/ast.hjson]: Updated!

**Comment by Jacob-Levy** [comment]
Please review, approve & merge...

**Comment by tjaychen** [comment]
sorry i haven't had time to get to this yet. Hopefully some time today or early next week...

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:53]: Suggest adding `%m` to all `$display` statements so that it is clear in the log, which module did the log message came from. 

```suggestion
  $display("%m: USB Clock Drift: %0d ps", rand32);
```

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:53]: What about doing a separate PR for adding the %m: in all the modules with $display?

Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv:53]: Yeah that's ok.

**Review by sriyerg — APPROVED** [review]

**Comment by sriyerg** [comment]
Will wait for @tjaychen & @matutem approval before merging.

File: hw/top_earlgrey/ip/ast/rtl/usb_osc.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/usb_osc.sv]: Add %m:  to all $display

**Comment by Jacob-Levy** [comment]
@tjaychen @matutem  Ready for your review

File: hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv:6]: I am not in favor of adding strange files and functions that seemingly dont do anything in the open source.

**Review by sriyerg — DISMISSED** [review]

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/data/ast.hjson:33]: This expansion seems quite unnecessary,

File: hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv:1]: A different plusarg method of handling this has been provided to you via email. Please try that method instead.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv:1]: This is loop unrolling to enable more precise exclude function to the AST registers. Some of the AST registers AST are status only registers and some are write port. Since we can't reveal the registers from security reason, I need to have a more presice exclude description of the registers.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv:1]: This is informational files.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:33]: This is loop unrolling to enable more precise exclude function to the AST registers. Some of the AST registers AST are status only registers and some are write port. Since we can't reveal the registers from security reason, I need to have a more presice exclude description of the registers.

File: hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv:6]: All the AST registers are external with special security features that breaks the RAL model. This file is an example on how to fix it.

File: hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv:1]: This is already well-documented here: 
https://github.com/lowRISC/opentitan/blob/4ac33d98b562fb25a55fb3d9793bce8e534134e5/hw/top_earlgrey/dv/env/chip_env_cfg.sv#L33

Please remove this file.

File: hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv:6]: You have your own internal `ast.hjson`. In your internal chip level DV env, you use this hjson to create the more accurate chip RAL which completely replaces the open source chip RAL. You are at the liberty of changing / fixing what you need to in your internal env. What is the need for this to exist in the open source?

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:33]: You have your own internal `ast.hjson` which you use to create the correct chip RAL, which replaces the open source chip RAL. When you add those exclusions in your internal `ast.hjson`, they will show up in your internal chip RAL. There is no need for this change in the open source, since the open source chip RAL and your internal chip RAL models are completely unrelated.

File: hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv:1]: @tjaychen @matutem 
Can you please review the rest of the files.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (sriyerg) [inline hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: Why not use the `fileset_partner` mechanism rather than introduce yet another mechanism to switch between open and closed source files? Use of macros like this should be avoided - You are basically counting on which file gets picked up first in the compilation order, so that the right set of parameters and types get defined, which is risky and hard to debug IMO. The `fileset_partner` mechanism already solves this by ensuring the right file that is picked and the wrong file does not even show up in the filelist used for compilation.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: This file is used for the OS' LFSR.

File: hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv:6]: During the review of the ast registers model we were asked to expose as much as we can on those registers.
I started doing that by unrolling the loop and updating the exclude information. Next will be the access information.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:33]: During the review of the ast registers model we were asked to expose as much as we can on those registers.
I started doing that by unrolling the loop and updating the exclude information. Next will be the access information.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: File updated to include only OS LFSR parameters information.

File: hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/backdoor_otp_ast_cfg.sv]: File removed for OS and no need for the plusarg method for now.

File: hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/dv/ast_base_reg_hdl_path.sv]: File removed from the OS.

**Comment by Jacob-Levy** [comment]
@sriyerg All issues you raised that I could fix were fixed/removed.
@tjaychen @matutem Other files are ready for your review.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:33]: Like I said, this RAL and the RAL generated by you in NTIL are completely separate. Anything you do here has no effect on your internal extensions. I am really not sure how this is going to be any useful.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: > As for the macro, it is for getting the content only once when the file is introduce more then one time.

Please see my first comment - my advice is exactly to not do this. The `fileset_partner` already is a mechanism put in place to pick one file or the other giong from open source to closed source - not pick both and rely on a macro define.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: @sriyerg It looks like that you might not understand the code or I don't understand your comment. Let set a meet/call to discuss.

**Comment by Jacob-Levy** [comment]
sw/device/tests/hmac_smoketest.c:11:10: fatal error: sw/device/lib/testing/hmac_testutils.h: No such file or directory
   11 | #include "sw/device/lib/testing/hmac_testutils.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.

FPGA code is broken...

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (tjaychen) [inline hw/top_earlgrey/ip/ast/data/ast.hjson:33]: hmm... i wonder if in the future it makes sense to allow for slight tweaks in multi-reg, ie they don't all have to be identical in attributes.  
@eunchan 
@msfschaffner 

what do you guys think? I think it boils down to what we think of mult-reg.
Are they a convenient array where we compact? or is it a duplication of a single register?

Allowing attribute changes could make things very complicated.. so i'm leaning towards no.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (tjaychen) [inline hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:15]: @Jacob-Levy @arnonsha 
will these seeds be different for proprietary AST? 

Bringing this up mostly because all the other random seeds in our design have an open source version, and another one that is generated for close source (Arnon has a script for this).  Does it make sense to merge this into that flow?

File: hw/top_earlgrey/ip/ast/rtl/ast_clks_byp.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_clks_byp.sv:424]: It does not complain on combinational logic going to a synchroniser as long it is from a single clock.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:15]: They are not identical. The proprietary one has more items in it.
I added this file to simplify  Arnon's script for the AST. Script already developed and tested by Arnon.

**Comment by arnonsha** [comment]
Yes, this will fit into the seed replacement flow.

File: hw/top_earlgrey/ip/ast/data/ast.hjson
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/data/ast.hjson:33]: I see no real value in that. If you need different attributes for each register, you better define each register separately.

For the real AST, all registers are external due to security features and as a result it breaks the automated RAL model that need to be fixed with special function (tried to share it...).

**Comment by Jacob-Levy** [comment]
@sriyerg Files was removed, please remove the change request.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (sriyerg) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: I understand your macro usage here very well. I am asking you to remove it because it should not even come to "prevention". We should not be introducing two separate files that contain the same set of parameters and types defined, and rely on the fact that the tool places the files in correct order, so that the "right" set is picked. IMO the file with the "wrong" set of parameters and types should not even appear in the filelist.

**Comment by sriyerg** [comment]
> @sriyerg Files were removed - Please remove the change request.

The point of contention now is the addition of ifdef guard in `ast_seeds_pkg` which needs to be resolved.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv:10]: This issue was discussed in the past and it was agreed to leave it as is.

Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv]: File removed frpm PR.

**Comment by Jacob-Levy** [comment]
@sriyerg File  `ast_seeds_pkg` removed from PR.

File: hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv
Comment (Jacob-Levy) [reply hw/top_earlgrey/ip/ast/rtl/ast_seed_pkg.sv]: File was remove from the OS and now exixt only in the clocse source.

**Comment by Jacob-Levy** [comment]
@sriyerg All issues you raised  were fixed/removed and your response "Will let someone else approve and merge".
@matutem Add flag `calibrate_usb_clk` to USB oscillator in this PR. 

@tjaychen @arnonsha 
This will be my last PR/Contribution to the Open Source,
Thank you all.

**Review by msfschaffner — APPROVED** [review]

**Pull request merged**
